pkg/root/trusted_root_create.go

```diff
@@ -0,0 +1,221 @@
+				iTime = tr.trustedRoot.Tlogs[i].PublicKey.ValidFor.Start.AsTime()
+			}
+			if tr.trustedRoot.Tlogs[j].PublicKey.ValidFor.Start != nil {
+				iTime = tr.trustedRoot.Tlogs[j].PublicKey.ValidFor.Start.AsTime()
+			}
+			return iTime.Before(jTime)
+		})
+	}
+
+	for logID, ctLog := range tr.ctLogs {
+		ctProto, err := transparencyLogToProtobufTL(*ctLog)
+		if err != nil {
+			return fmt.Errorf("failed converting ctlog %s to protobuf: %w", logID, err)
+		}
+		tr.trustedRoot.Ctlogs = append(tr.trustedRoot.Ctlogs, ctProto)
+		// ensure stable sorting of the slice
+		sort.Slice(tr.trustedRoot.Ctlogs, func(i, j int) bool {
+			iTime := time.Unix(0, 0)
+			jTime := time.Unix(0, 0)
+
+			if tr.trustedRoot.Ctlogs[i].PublicKey.ValidFor.Start != nil {
+				iTime = tr.trustedRoot.Ctlogs[i].PublicKey.ValidFor.Start.AsTime()
+			}
+			if tr.trustedRoot.Ctlogs[j].PublicKey.ValidFor.Start != nil {
+				iTime = tr.trustedRoot.Ctlogs[j].PublicKey.ValidFor.Start.AsTime()
+			}
+			return iTime.Before(jTime)
+		})
+	}
+
+	for _, ca := range tr.fulcioCertAuthorities {
+		caProto, err := certificateAuthorityToProtobufCA(ca)
+		if err != nil {
+			return fmt.Errorf("failed converting fulcio cert chain to protobuf: %w", err)
+		}
+		tr.trustedRoot.CertificateAuthorities = append(tr.trustedRoot.CertificateAuthorities, caProto)
+	}
+
+	for _, ca := range tr.timestampingAuthorities {
+		caProto, err := certificateAuthorityToProtobufCA(ca)
+		if err != nil {
+			return fmt.Errorf("failed converting TSA cert chain to protobuf: %w", err)
+		}
+		tr.trustedRoot.TimestampAuthorities = append(tr.trustedRoot.TimestampAuthorities, caProto)
+	}
+
+	return nil
+}
+
+func certificateAuthorityToProtobufCA(ca CertificateAuthority) (*prototrustroot.CertificateAuthority, error) {
+	org := ""
+	if len(ca.Root.Subject.Organization) > 0 {
+		org = ca.Root.Subject.Organization[0]
+	}
+	var allCerts []*protocommon.X509Certificate
+	if ca.Leaf != nil {
+		allCerts = append(allCerts, &protocommon.X509Certificate{RawBytes: ca.Leaf.Raw})
+	}
+	for _, intermed := range ca.Intermediates {
+		allCerts = append(allCerts, &protocommon.X509Certificate{RawBytes: intermed.Raw})
+	}
+	if ca.Root == nil {
+		return nil, fmt.Errorf("root certificate is nil")
+	}
+	allCerts = append(allCerts, &protocommon.X509Certificate{RawBytes: ca.Root.Raw})
+
+	caProto := prototrustroot.CertificateAuthority{
+		Uri: ca.URI,
+		Subject: &protocommon.DistinguishedName{
+			Organization: org,
+			CommonName:   ca.Root.Subject.CommonName,
+		},
+		ValidFor: &protocommon.TimeRange{
+			Start: timestamppb.New(ca.ValidityPeriodStart),
+		},
+		CertChain: &protocommon.X509CertificateChain{
+			Certificates: allCerts,
+		},
+	}
+
+	if !ca.ValidityPeriodEnd.IsZero() {
+		caProto.ValidFor.End = timestamppb.New(ca.ValidityPeriodEnd)
+	}
+
+	return &caProto, nil
+}
+
+func transparencyLogToProtobufTL(tl TransparencyLog) (*prototrustroot.TransparencyLogInstance, error) {
```

Comment:
```suggestion
func transparencyLogToProtobufTL(tl *TransparencyLog) (*prototrustroot.TransparencyLogInstance, error) {
```

Any reason not to receive this via a pointer to avoide excessive copying?

Reply:
You're absolutely right, I missed that. I will fix it.